File: v21_proxy/python/image_proxy.py
class RemoteImage:
    def __init__(self, url):
        self.url = url

    def display_image(self):
        print("Displaying image from URL:", self.url)
        # Code to load and display the image


class ImageProxy:
    def __init__(self, url):
        self.url = url
        self.image = None

    def display_image(self):
        if self.image is None:
            print("New image from URL:", self.url)
            self.image = RemoteImage(self.url)
        self.image.display_image()


def main():
    image_proxy = ImageProxy("http://example.com/image.jpg")
    image_proxy.display_image()  # Image is loaded from the server
    image_proxy.display_image()  # Image is loaded from the cache


if __name__ == "__main__":
    main()

File: v21_proxy/java/ImageProxy.java
package v21_proxy.java;

class  RemoteImage {
    private String url;

    public RemoteImage(String url) {
        this.url = url;
    }

    public void displayImage() {
        System.out.println("Displaying image from URL: " + url);
        // Code to load and display the image
    }
}

public class ImageProxy {
    private String url;
    private RemoteImage image;

    public ImageProxy(String url) {
        this.url = url;
        this.image = null;
    }

    public void displayImage() {
        if (image == null) {
            System.out.println("New image from URL: " + url);
            image = new RemoteImage(url);
        }
        image.displayImage();
    }

    public static void main(String[] args) {
        ImageProxy imageProxy = new ImageProxy("http://example.com/image.jpg");
        imageProxy.displayImage(); // Image is loaded from the server
        imageProxy.displayImage(); // Image is loaded from the cache
    }
}
